browser/ipfs/ipfs_tab_helper.cc

```diff
@@ -62,19 +111,115 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
+    return;
+  std::string dnslink = GetDNSRecordValue(text_results, kDnsLinkHeader);
+  if (dnslink.empty())
+    return;
+  ipfs_resolved_host_ = host;
+  if (pref_service_->GetBoolean(kIPFSAutoRedirectDNSLink)) {
+    content::OpenURLParams params(GetIPFSResolvedURL(), content::Referrer(),
+                                  WindowOpenDisposition::CURRENT_TAB,
+                                  ui::PAGE_TRANSITION_LINK, false);
+    web_contents()->OpenURL(params);
+    return;
+  }
+  UpdateLocationBar();
+}
+
+void IPFSTabHelper::UpdateLocationBar() {
+  if (web_contents()->GetDelegate())
+    web_contents()->GetDelegate()->NavigationStateChanged(
+        web_contents(), content::INVALIDATE_TYPE_URL);
+}
+
+GURL IPFSTabHelper::GetIPFSResolvedURL() const {
+  if (ipfs_resolved_host_.empty())
+    return GURL();
+  GURL current = web_contents()->GetURL();
+  GURL::Replacements replacements;
+  replacements.SetSchemeStr(kIPNSScheme);
+  return current.ReplaceComponents(replacements);
+}
+
+bool IPFSTabHelper::HasActiveResolverForHost(const std::string& host) const {
+  for (auto& resolver : resolvers_) {
+    if (!resolver->IsRunning())
+      continue;
+    if (resolver->host() == host)
+      return true;
+  }
+  return false;
+}
+
+void IPFSTabHelper::ResolveIPFSLink() {
+  GURL current = web_contents()->GetURL();
+  if (!current.SchemeIsHTTPOrHTTPS()) {
+    return;
+  }
+  auto resolve_method = static_cast<ipfs::IPFSResolveMethodTypes>(
+      pref_service_->GetInteger(kIPFSResolveMethod));
+  if (resolve_method == ipfs::IPFSResolveMethodTypes::IPFS_DISABLED ||
+      resolve_method == ipfs::IPFSResolveMethodTypes::IPFS_ASK)
+    return;
+
+  if (ipfs_resolved_host_ == current.host())
+    return;
+
+  if (HasActiveResolverForHost(current.host()))
+    return;
+
+  const auto& host_port_pair = net::HostPortPair::FromURL(current);
+  for (auto& resolver : resolvers_) {
+    if (resolver->IsRunning())
+      resolver->Reset();
+
+    auto resolved_callback = base::BindOnce(
+        &IPFSTabHelper::HostResolvedCallback, weak_ptr_factory_.GetWeakPtr());
+    const auto& key =
+        web_contents()->GetMainFrame()
+            ? web_contents()->GetMainFrame()->GetNetworkIsolationKey()
+            : net::NetworkIsolationKey();
+    resolver->Resolve(host_port_pair, key, net::DnsQueryType::TXT,
+                      std::move(resolved_callback));
+  }
+}
+
 void IPFSTabHelper::DidFinishNavigation(content::NavigationHandle* handle) {
   DCHECK(handle);
-  if (!handle->IsInMainFrame()) {
+  if (!handle->IsInMainFrame() || !handle->HasCommitted() ||
+      !handle->GetResponseHeaders()) {
+    return;
+  }
+
+  if (!handle->IsSameDocument() && !ipfs_resolved_host_.empty()) {
+    ipfs_resolved_host_.erase();
+    UpdateLocationBar();
+  }
+
+  int response_code = handle->GetResponseHeaders()->response_code();
+  if (ipfs_resolved_host_.empty() &&
+      response_code >= net::HttpStatusCode::HTTP_INTERNAL_SERVER_ERROR &&
```

Comment:
can you also add comments here? Are there any ipfs docs you can link to explaining why we're handling this subset of status codes?

Reply:
https://github.com/brave/brave-browser/issues/13609#issuecomment-764627805
```
or if page returned code different from OK 200 or failed due to network/DNS (eg. invalid A record)
(this will make it possible for people to recover from broken/DDoSs HTTP servers etc)
```